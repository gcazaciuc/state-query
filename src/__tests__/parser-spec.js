const nearley = require('nearley');
const grammar = require('../../grammar.js');

describe('State Query SQL parser suite', () => {
    test('Should parse simple SQL statements', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('SELECT col FROM table');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
    });

    test('Should parse simple SQL statements when keyword are lower-case', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('select col from table');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
        // console.log(JSON.stringify(r));
    });

    test('Should parse SQL statements with where clauses in it - both members are identifiers', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('select col from table where col >x');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
        // console.log(JSON.stringify(r));
    });

    test('Should parse SQL statements with where clauses in it - one member is a number literal', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('select col from table where col>5');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
        // console.log(JSON.stringify(r));
    });

    test('Should parse SQL statements with existential operator in it', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('select * from table');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
        // console.log(JSON.stringify(r));
    });

    test('Should parse SQL create table statements', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('create table users(userId int)');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
        // console.log(JSON.stringify(r));
    });

    test('Should parse insert into table statements', () => {
        const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
        parser.feed('insert into users(name) values(active, no)');
        const r = parser.results[0];
        expect(r).toMatchSnapshot();
        console.log(JSON.stringify(r));
    });
});