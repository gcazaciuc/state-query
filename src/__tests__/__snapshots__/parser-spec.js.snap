// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State Query SQL parser suite Should parse SQL create table statements 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "alias": "userId",
        "data": Array [
          "int",
        ],
        "name": "userId",
        "type": "col",
      },
      null,
    ],
    "name": Array [
      "users",
    ],
    "type": "create_table",
  },
]
`;

exports[`State Query SQL parser suite Should parse SQL statements with existential operator in it 1`] = `
Array [
  Object {
    "clauses": null,
    "cols": Object {
      "cols": Array [
        "*",
      ],
      "type": "columnList",
    },
    "from": Object {
      "tables": Array [
        "table",
      ],
      "type": "from",
    },
    "type": "select",
  },
]
`;

exports[`State Query SQL parser suite Should parse SQL statements with where clauses in it - both members are identifiers 1`] = `
Array [
  Object {
    "clauses": Array [
      Array [
        Object {
          "clauses": Object {
            "left": "col",
            "operator": ">",
            "right": "x",
            "type": "binaryExpression",
          },
          "type": "where",
        },
      ],
    ],
    "cols": Object {
      "cols": Array [
        Object {
          "alias": "col",
          "name": "col",
          "type": "column",
        },
      ],
      "type": "columnList",
    },
    "from": Object {
      "tables": Array [
        "table",
      ],
      "type": "from",
    },
    "type": "select",
  },
]
`;

exports[`State Query SQL parser suite Should parse SQL statements with where clauses in it - one member is a number literal 1`] = `
Array [
  Object {
    "clauses": Array [
      Array [
        Object {
          "clauses": Object {
            "left": "col",
            "operator": ">",
            "right": Array [
              Array [
                "5",
              ],
            ],
            "type": "binaryExpression",
          },
          "type": "where",
        },
      ],
    ],
    "cols": Object {
      "cols": Array [
        Object {
          "alias": "col",
          "name": "col",
          "type": "column",
        },
      ],
      "type": "columnList",
    },
    "from": Object {
      "tables": Array [
        "table",
      ],
      "type": "from",
    },
    "type": "select",
  },
]
`;

exports[`State Query SQL parser suite Should parse simple SQL statements 1`] = `
Array [
  Object {
    "clauses": null,
    "cols": Object {
      "cols": Array [
        Object {
          "alias": "col",
          "name": "col",
          "type": "column",
        },
      ],
      "type": "columnList",
    },
    "from": Object {
      "tables": Array [
        "table",
      ],
      "type": "from",
    },
    "type": "select",
  },
]
`;

exports[`State Query SQL parser suite Should parse simple SQL statements when keyword are lower-case 1`] = `
Array [
  Object {
    "clauses": null,
    "cols": Object {
      "cols": Array [
        Object {
          "alias": "col",
          "name": "col",
          "type": "column",
        },
      ],
      "type": "columnList",
    },
    "from": Object {
      "tables": Array [
        "table",
      ],
      "type": "from",
    },
    "type": "select",
  },
]
`;
